File: server/src/models/auth/Session.ts
import * as uuid from "https://deno.land/std@0.175.0/uuid/mod.ts";

import "./Capability.ts"
import { Capabilities } from "./Capability.ts";

export class AccessToken {
    rawToken: string;

    constructor(raw: string) {
        this.rawToken = raw;
    }

    static generate(): AccessToken {
        return new AccessToken(uuid.v4.generate())
    }
}

export class UserId {
    rawId: string;

    constructor(raw: string) {
        this.rawId = raw;
    }

    static generate(): UserId {
        return new UserId(uuid.v4.generate())
    }
}

export enum Role {
    Admin,
    Reviewer,
    Publisher,
}

export class UserSession {
    user_id:        UserId;
    access_token:   AccessToken;
    expiration:     Date;
    permissions:    Capabilities;
    role:           Role;

    constructor(
        user_id:        UserId,
        access_token:   AccessToken,
        expiration:     Date,
        permissions:    Capabilities,
        role:           Role)
    {
        this.user_id = user_id;
        this.access_token = access_token;
        this.expiration = expiration;
        this.permissions = permissions;
        this.role = role;
    }
}

File: server/src/models/session.rs
use postgres_types::{FromSql, ToSql};
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Local};
use uuid::Uuid;

uuid_based! (AccessToken);
uuid_based! (InvalidToken);
uuid_based! (UserId);

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum Role {
    Admin,
    Publisher,
    User,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct UserSession {
    pub user_id:        UserId,
    pub access_token:   AccessToken,
    pub expiration:     DateTime<Local>,
    // capabilities
    pub role:           Role,
}

impl Serialize for Role {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer {
        match self {
            Role::Admin => serializer.serialize_str("admin"),
            Role::Publisher => serializer.serialize_str("publisher"),
            Role::User => serializer.serialize_str("user"),
        }
    }
}

impl<'de> Deserialize<'de> for Role {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>
    {
        let str_role = String::deserialize(deserializer)?;

        let role =
        match str_role.as_str() {
            "admin" => Role::Admin,
            "publisher" => Role::Publisher,
            "user" => Role::User,
            _ => {
                return Err(serde::de::Error::custom(
                    "invalid role",
                ));
            }
        };

        Ok(role)
    }
}

impl ToSql for Role {
    fn to_sql(&self, ty: &postgres_types::Type, out: &mut postgres_types::private::BytesMut) -> Result<postgres_types::IsNull, Box<dyn std::error::Error + Sync + Send>>
        where Self: Sized
    {
        match self {
            Role::Admin => "admin".to_sql(ty, out),
            Role::Publisher => "publisher".to_sql(ty, out),
            Role::User => "user".to_sql(ty, out),
        }
    }

    fn accepts(ty: &postgres_types::Type) -> bool
        where Self: Sized
    {
        <String as ToSql>::accepts(ty)
    }

    fn to_sql_checked(
        &self,
        ty: &postgres_types::Type,
        out: &mut postgres_types::private::BytesMut,
    ) -> Result<postgres_types::IsNull, Box<dyn std::error::Error + Sync + Send>> {
        match self {
            Role::Admin => "admin".to_sql_checked(ty, out),
            Role::Publisher => "publisher".to_sql_checked(ty, out),
            Role::User => "user".to_sql_checked(ty, out),
        }
    }
}

impl FromSql<'_> for Role {
    fn from_sql(ty: &postgres_types::Type, raw: &[u8]) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
        let str_role = String::from_sql(ty, raw)?;

        let role =
        match str_role.as_str() {
            "admin" => Role::Admin,
            "publisher" => Role::Publisher,
            "user" => Role::User,
            _ => {
                // todo: just return an error
                panic!("corrupt table");
            }
        };

        Ok(role)
    }

    fn accepts(ty: &postgres_types::Type) -> bool {
        <String as FromSql>::accepts(ty)
    }
}